### PR TITLE
added skybox primitive

### PR DESCRIPTION
**Pull request by phaseOne**
**Description:**
Addresses #45.

Still have issue that the cubemap is 180 flipped in y in relation to the viewer inside the cubemap. The default shader does not have a means of rotation.

### CODE REVIEW

**Comment by ngokevin** [comment]
Thanks for this, will look when we can. Got a bit of a backlog :)

**Comment by ngokevin** [comment]
This looks good! Can you add a quick and cool example in `test/cubemaps`?

**Comment by ngokevin** [comment]
Going to mark needs changes to add an example

File: src/shaders/cube.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  /**
+   * Updating existing material.
+   *
+   * @param {object} data - Material component data.
+   * @returns {object} Material.
+   */
+  updateMaterial: function (data) {
+    var material = this.material;
+    data = getMaterialData(data);
+    Object.keys(data).forEach(function (key) {
```
Comment (ngokevin) [inline src/shaders/cube.js]: can name this function like `setMaterialData`

File: src/shaders/cube.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+var CubeLoader = new THREE.CubeTextureLoader();
+var texturePromises = {};
+
+/**
+ * Cubemap shader using THREE.ShaderMaterial.
+ */
+module.exports.Component = registerShader('cube', {
+  schema: {
+    src: { default: '' }
```
Comment (ngokevin) [inline src/shaders/cube.js]: let's make no spaces: `{default: ''}`

File: src/shaders/cube.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  },
+  /**
+   * Initializes the shader.
+   * Adds a reference from the scene to this entity as the camera.
+   */
+  init: function (data) {
+    this.material = new THREE.ShaderMaterial(getMaterialData(data));
+    this.updateCubeMap(data);
+    return this.material;
```
Comment (ngokevin) [inline src/shaders/cube.js]: I just updated to not have to return this :) https://github.com/aframevr/aframe/pull/1549

**Comment by cvan** [comment]
@phaseOne: This is awesome work here. It would be awesome to replace a scene such as [this one](https://codepen.io/vinsongrant/pen/grzOmb?editors=1000#0) with more succinct markup enabled by this `<a-skybox>` primative. Let us know if you need any help. 😄

File: src/shaders/cube.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+   * Handle texture cubemap. Textures are cached in texturePromises.
+   */
+  updateCubeMap: function (data) {
+    var self = this;
+    var material = this.material;
+    var cubeMap = data.src;
+
+    // No cubeMap defined or already loading.
+    if (!cubeMap || this.isLoadingCubeMap) {
+      material.uniforms['tCube'].value = null;
```
Comment (cvan) [inline src/shaders/cube.js]: dot-key syntax would be fine here:

```js
material.uniforms.tCube.value = null;
```

File: src/shaders/cube.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      material.needsUpdate = true;
+      return;
+    }
+    this.isLoadingCubeMap = true;
+
+    // Another material is already loading this texture. Wait on promise.
+    if (texturePromises[cubeMap]) {
+      texturePromises[cubeMap].then(function (cube) {
+        self.isLoadingCubeMap = false;
+        material.uniforms['tCube'].value = cube;
```
Comment (cvan) [inline src/shaders/cube.js]: same here: dot-key syntax would be fine here:

```js
material.uniforms.tCube.value = cube;
```

File: src/shaders/cube.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      return;
+    }
+
+    // Material is first to load this texture. Load and resolve texture.
+    texturePromises[cubeMap] = new Promise(function (resolve) {
+      utils.srcLoader.validateCubemapSrc(cubeMap, function loadCubeMap (urls) {
+        CubeLoader.load(urls, function (cube) {
+          // Texture loaded.
+          self.isLoadingCubeMap = false;
+          material.uniforms['tCube'].value = cube;
```
Comment (cvan) [inline src/shaders/cube.js]: same here:

dot-key syntax would be fine here:

```js
material.uniforms.tCube.value = cube;
```

File: src/shaders/cube.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+          resolve(cube);
```
Comment (cvan) [inline src/shaders/cube.js]: any errors to reject upon from `utils.srcLoader.validateCubemapSrc` and/or `CubeLoader.load`?

File: src/shaders/cube.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+});
+
+/**
+ * Builds and normalize material data, normalizing stuff along the way.
+ *
+ * @param {object} data - Material data.
+ * @returns {object} data - Processed material data.
+ */
+function getMaterialData (data) {
+  var cubeShader = THREE.ShaderLib['cube'];
```
Comment (cvan) [inline src/shaders/cube.js]: this can be `THREE.ShaderLib.cube`

File: src/shaders/cube.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+var texturePromises = {};
+
+/**
+ * Cubemap shader using THREE.ShaderMaterial.
+ */
+module.exports.Component = registerShader('cube', {
+  schema: {
+    src: { default: '' }
+  },
+  /**
```
Comment (cvan) [inline src/shaders/cube.js]: to be consistent, you could add a newline above

File: src/extras/primitives/primitives/a-skybox.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+var getMeshMixin = require('../getMeshMixin');
+var registerPrimitive = require('../registerPrimitive');
+var utils = require('../../../utils/');
+
+registerPrimitive('a-skybox', utils.extendDeep({}, getMeshMixin(), {
```
Comment (cvan) [inline src/extras/primitives/primitives/a-skybox.js]: feel free to file a follow-up issue for it instead, but it'd be awesome to at least stub out the reference Docs page for this primitive at `docs/primitives/a-skybox.md` (see [`docs/primitives/a-sky.md`](https://github.com/aframevr/aframe/blob/master/docs/primitives/a-sky.md) for an example).

**Comment by cvan** [comment]
@phaseOne did you want to address the remaining issues so this can get merged?

**Comment by cvan** [comment]
how's this PR look, @dmarcos, @ngokevin?

**Pull request closed without merging**